modules/xfem/src/userobjects/GeometricCut2DUserObject.C

```diff
@@ -0,0 +1,227 @@
+                                        seg_int_frac))
+        {
+          cut_frag = true;
+          CutEdge mycut;
+          mycut.id1 = i;
+          mycut.id2 = (i < (n_sides - 1) ? (i + 1) : 0);
+          mycut.distance = seg_int_frac;
+          mycut.host_side_id = i;
+          cut_edges.push_back(mycut);
+        }
+      }
+    }
+  }
+  return cut_frag;
+}
+
+bool
+GeometricCut2DUserObject::cutFragmentByGeometry(std::vector<std::vector<Point>> & /*frag_faces*/,
+                                                std::vector<CutFace> & /*cut_faces*/,
+                                                Real /*time*/) const
+{
+  mooseError("Invalid method: must use vector of element edges for 2D mesh cutting");
+  return false;
+}
+
+bool
+GeometricCut2DUserObject::IntersectSegmentWithCutLine(const Point & segment_point1,
+                                                      const Point & segment_point2,
+                                                      const Point & cutting_line_point1,
+                                                      const Point & cutting_line_point2,
+                                                      const Real & cutting_line_fraction,
+                                                      Real & segment_intersection_fraction) const
+{
+  // Use the algorithm described here to determine whether a line segment is intersected
+  // by a cutting line, and to compute the fraction along that line where the intersection
+  // occurs:
+  // http://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
+
+  bool cut_segment = false;
+  Point seg_dir = segment_point2 - segment_point1;
+  Point cut_dir = cutting_line_point2 - cutting_line_point1;
+  Point cut_start_to_seg_start = segment_point1 - cutting_line_point1;
+
+  Real cut_dir_cross_seg_dir = crossProduct2D(cut_dir(0), cut_dir(1), seg_dir(0), seg_dir(1));
+
+  if (std::abs(cut_dir_cross_seg_dir) > Xfem::tol)
+  {
+    // Fraction of the distance along the cutting segment where it intersects the edge segment
+    Real cut_int_frac =
+        crossProduct2D(
+            cut_start_to_seg_start(0), cut_start_to_seg_start(1), seg_dir(0), seg_dir(1)) /
+        cut_dir_cross_seg_dir;
+
+    if (cut_int_frac >= 0.0 && cut_int_frac <= cutting_line_fraction)
+    { // Cutting segment intersects the line of the edge segment, but the intersection point may be
+      // outside the segment
+      Real int_frac =
+          crossProduct2D(
+              cut_start_to_seg_start(0), cut_start_to_seg_start(1), cut_dir(0), cut_dir(1)) /
+          cut_dir_cross_seg_dir;
+      if (int_frac >= 0.0 &&
+          int_frac <= 1.0) // TODO: revisit end cases for intersections with corners
+      {
+        cut_segment = true;
+        segment_intersection_fraction = int_frac;
+      }
+    }
+  }
+  return cut_segment;
+}
+
+Real
+GeometricCut2DUserObject::crossProduct2D(Real ax, Real ay, Real bx, Real by) const
```

Comment:
would be more elegant if this function took two `const Point &` parameters (to shift the cluttter from the caller side into the function)